core/dbt/deps/tarball.py

```diff
@@ -42,23 +48,31 @@ def nice_version_name(self):
         return f"tarball (url: {self.tarball})"
 
     def _fetch_metadata(self, project, renderer):
-        """
-        recycle RegistryPackageMetadata so that we can use the install and
-        download_and_untar from RegistryPinnedPackage next.
-        build RegistryPackageMetadata from info passed via packages.yml since no
-        'metadata' service exists in this case.
-        """
-
-        dct = {
-            "name": self.package,
-            "packages": [],  # note: required by RegistryPackageMetadata
-            "downloads": {"tarball": self.tarball},
-        }
-
-        return RegistryPackageMetadata.from_dict(dct)
+        """Download and untar the project and parse metadata from the project folder."""
+        system.download(self.tarball, self.tar_path)
+        system.untar_package(self.tar_path,self.untarred_path)
+        tar_contents = os.listdir(self.untarred_path)
+        if len(tar_contents) != 1:
+            raise DependencyError(
+                f"Incorrect structure for package extracted from {self.tarball}."
+                f"The extracted package needs to follow the structure {self.name}/<package_contents>."
+            )
+        child_folder = os.listdir(self.untarred_path)[0]
+
+        self.untarred_path = os.path.join(self.untarred_path,child_folder)
+        partial = PartialProject.from_project_root(self.untarred_path)
+        metadata = partial.render_package_metadata(renderer)
+        metadata.name = self.package if self.package else metadata.name
+        return metadata
 
     def install(self, project, renderer):
-        self._install(project, renderer)
+        dest_path = self.get_installation_path(project, renderer)
+        if os.path.exists(dest_path):
+            if system.path_is_symlink(dest_path):
+                system.remove_file(dest_path)
+            else:
+                system.rmdir(dest_path)
+        system.move(self.untarred_path, dest_path)
```

Comment:
This assumes that `_fetch_metadata` has always already been run. That might not be the case with the changes to `dbt deps` in v1.7 — the introduction of `package-lock.json` — which separates these two steps:
1. Resolving package metadata (to generate/update the lock file)
2. Installing package contents (from the lock file specification)